src/ca/mcgill/cs/jetuml/gui/GraphPanel2.java

```diff
@@ -0,0 +1,992 @@
+ *******************************************************************************/
+
+package ca.mcgill.cs.jetuml.gui;
+
+import java.util.HashSet;
+import java.util.ResourceBundle;
+import java.util.Set;
+import java.util.Stack;
+
+import ca.mcgill.cs.jetuml.application.Clipboard2;
+import ca.mcgill.cs.jetuml.application.GraphModificationListener2;
+import ca.mcgill.cs.jetuml.application.MoveTracker2;
+import ca.mcgill.cs.jetuml.application.PropertyChangeTracker;
+import ca.mcgill.cs.jetuml.application.SelectionList;
+import ca.mcgill.cs.jetuml.application.UndoManager;
+import ca.mcgill.cs.jetuml.commands.AddEdgeCommand2;
+import ca.mcgill.cs.jetuml.commands.AddNodeCommand2;
+import ca.mcgill.cs.jetuml.commands.ChangePropertyCommand;
+import ca.mcgill.cs.jetuml.commands.CompoundCommand;
+import ca.mcgill.cs.jetuml.commands.DeleteNodeCommand2;
+import ca.mcgill.cs.jetuml.commands.RemoveEdgeCommand2;
+import ca.mcgill.cs.jetuml.geom.Line;
+import ca.mcgill.cs.jetuml.geom.Point;
+import ca.mcgill.cs.jetuml.geom.Rectangle;
+import ca.mcgill.cs.jetuml.geom.Zoom;
+import ca.mcgill.cs.jetuml.graph.Edge;
+import ca.mcgill.cs.jetuml.graph.Graph2;
+import ca.mcgill.cs.jetuml.graph.GraphElement;
+import ca.mcgill.cs.jetuml.graph.Node;
+import ca.mcgill.cs.jetuml.graph.Property;
+import ca.mcgill.cs.jetuml.graph.nodes.ChildNode;
+import ca.mcgill.cs.jetuml.graph.nodes.ImplicitParameterNode;
+import ca.mcgill.cs.jetuml.graph.nodes.ObjectNode;
+import ca.mcgill.cs.jetuml.graph.nodes.PackageNode;
+import ca.mcgill.cs.jetuml.graph.nodes.ParentNode;
+import ca.mcgill.cs.jetuml.views.Grid2;
+import javafx.event.EventHandler;
+import javafx.geometry.Bounds;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.geometry.Rectangle2D;
+import javafx.scene.Parent;
+import javafx.scene.Scene;
+import javafx.scene.canvas.Canvas;
+import javafx.scene.canvas.GraphicsContext;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.image.Image;
+import javafx.scene.input.MouseEvent;
+import javafx.scene.layout.BorderPane;
+import javafx.scene.paint.Color;
+import javafx.scene.paint.Paint;
+import javafx.stage.Modality;
+import javafx.stage.Stage;
+
+/**
+ * A panel to draw a graph.
+ * @author Kaylee I. Kutschera - Migration to JavaFX
+ */
+public class GraphPanel2 extends Canvas
+{
+	private enum DragMode 
+	{ DRAG_NONE, DRAG_MOVE, DRAG_RUBBERBAND, DRAG_LASSO }
+	
+	private static final int CONNECT_THRESHOLD = 8;
+	private static final int LAYOUT_PADDING = 20;
+	private static final Color GRABBER_COLOR = Color.rgb(77, 115, 153);
+	private static final Color GRABBER_FILL_COLOR = Color.rgb(173, 193, 214);
+	private static final Color GRABBER_FILL_COLOR_TRANSPARENT = Color.rgb(173, 193, 214, 0.75);
+	private static final int CANVAS_SCREEN_FACTOR = 2;	// factor to compute maximum canvas size
+	
+	private GraphicsContext aGraphics;
+	private GraphFrame2 aFrame;

```

Comment:
This should now be accessible through the component hierarchy. Please remove and replace with a private accessor method, as this is a great opportunity to reduce managed state in this object.

Reply:
Completed.